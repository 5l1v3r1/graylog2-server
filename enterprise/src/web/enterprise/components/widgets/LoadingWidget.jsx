import React from 'react';
import PropTypes from 'prop-types';

import styles from './LoadingWidget.css';

const LoadingWidget = () => (
  <div className={styles.spinnerContainer}>
    <i className="fa fa-spin fa-3x fa-refresh spinner" />
  </div>
);

LoadingWidget.propTypes = {};

export default LoadingWidget;
