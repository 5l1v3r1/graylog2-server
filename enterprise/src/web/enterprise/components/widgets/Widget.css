:local(.widgetActionDropdownCaret) {
    cursor: pointer;
    opacity: 0.3;
    font-size: 14px !important;
}
