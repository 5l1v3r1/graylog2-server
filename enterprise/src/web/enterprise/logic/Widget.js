import uuid from 'uuid/v4';

import { PluginStore } from 'graylog-web-plugin/plugin';

import { TIMESTAMP_FIELD } from 'enterprise/Constants';
import pivotForField from './searchtypes/aggregation/PivotGenerator';
import AggregationWidget from './aggregationbuilder/AggregationWidget';
import AggregationWidgetConfig from './aggregationbuilder/AggregationWidgetConfig';
import Series from './aggregationbuilder/Series';
import FieldType from './fieldtypes/FieldType';

const widgetsKey = 'enterpriseWidgets';

export function widgetDefinition(type) {
  return PluginStore.exports(widgetsKey)
    .find(widget => widget.type.toLocaleUpperCase() === type.toLocaleUpperCase());
}

export const resultHistogram = (id = uuid()) => AggregationWidget.builder()
  .id(id)
  .config(
    AggregationWidgetConfig.builder()
      .columnPivots([])
      .rowPivots([
        pivotForField(TIMESTAMP_FIELD, new FieldType('date', [], [])),
      ])
      .series([
        Series.forFunction('count()'),
      ])
      .sort([])
      .visualization('bar')
      .rollup(true)
      .build(),
  )
  .build();
