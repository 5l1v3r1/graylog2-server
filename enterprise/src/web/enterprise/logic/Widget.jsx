import { PluginStore } from 'graylog-web-plugin/plugin';

const widgetsKey = 'enterpriseWidgets';

export function widgetDefinition(type) {
  return PluginStore.exports(widgetsKey)
    .find(widget => widget.type.toLocaleUpperCase() === type.toLocaleUpperCase())
}

export const messageList = (id, timeRange, fields) => {
  return {
    id: id,
    type: 'messages',
    title: 'Messages',
    config: {
      fields: fields,
      showMessageRow: true,
    },
    computationTimeRange: timeRange,
  };
};

export const resultHistogram = (id, timeRange = {}) => {
  return {
    id: id,
    type: 'SEARCH_RESULT_CHART2',
    title: 'Histogram',
    computationTimeRange: timeRange,
    config: {
      timerange: timeRange,
    },
  };
};

export const searchSideBar = () => {
  return {
    type: 'SEARCH_SIDEBAR',
    title: 'Search Result',
    config: {},
    data: 'messages',
  };
};

export const dataTable = (id) => {
  return {
    id,
    type: 'DATATABLE',
    title: 'Results',
    config: {
      fields: ['action', 'controller', 'count'],
      data: [
        { action: 'index', controller: 'PostsController', count: 19016 },
        { controller: 'UsersController', count: 1376 },
        { action: 'login', controller: 'LoginController', count: 5156 },
        { action: 'show', controller: 'PostsController', count: 5903 },
        { action: 'edit', controller: 'PostsController', count: 922 },
      ],
    },
  };
};

const createWidgetDefinitions = (timeRange, fields) => {
  return {
    data: {},
    widgets: {
      messages: messageList(timeRange, fields),
      histogram: resultHistogram(timeRange),
      searchSideBar: searchSideBar(),
    },
    positions: {
      searchSideBar: { row: 1, col: 1, width: 1, height: 3.2 },
      histogram: { row: 1, col: 2, width: 4, height: 1.2 },
      messages: { row: 255, col: 2, width: 4, height: 3 },
    },
  };
};

