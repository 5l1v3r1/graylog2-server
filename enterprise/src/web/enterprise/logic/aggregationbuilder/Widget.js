import Immutable from 'immutable';

export default class Widget {
  constructor(id, type, config) {
    this._value = { id, type, config };
  }

  get id() {
    return this._value.id;
  }

  get type() {
    return this._value.type;
  }

  get config() {
    return this._value.config;
  }

  duplicate(newId) {
    return new Widget(newId, this.type, this.config);
  }

  toBuilder() {
    const { id, type, config } = this._value;
    return new Builder(Immutable.Map({ id, type, config }));
  }

  toJSON() {
    const { id, type, config } = this._value;

    return { id, type: type.toLocaleLowerCase(), config };
  }

  static fromJSON(value) {
    const { id, type, config } = value;
    return new Widget(id, type, Object.assign({}, config));
  }
}

class Builder {
  constructor(value = Immutable.Map()) {
    this.value = value;
  }

  id(value) {
    return new Builder(this.value.set('id', value));
  }

  type(value) {
    return new Builder(this.value.set('type', value));
  }

  config(value) {
    return new Builder(this.value.set('config', value));
  }

  build() {
    const { id, type, config } = this.value.toObject();
    return new Widget(id, type, config);
  }
}
