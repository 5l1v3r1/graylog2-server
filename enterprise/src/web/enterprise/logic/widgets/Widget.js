// @flow
import { Map } from 'immutable';
import uuid from 'uuid/v4';

type State = {
  id: string,
  type: string,
  config: any,
  filter: string;
};

class Widget {
  _value: State;
  constructor(id: string, type: string, config: any, filter: string) {
    this._value = { id, type, config, filter: filter === null ? undefined : filter };
  }

  static __registrations = {};

  get id() : string {
    return this._value.id;
  }

  get type() : string {
    return this._value.type;
  }

  get config() {
    return this._value.config;
  }

  get filter() : string {
    return this._value.filter;
  }

  duplicate(newId: string) : Widget {
    return this.toBuilder().id(newId).build();
  }

  toBuilder() : Builder {
    const { id, type, config, filter } = this._value;
    // eslint-disable-next-line no-use-before-define
    return new Builder(Map({ id, type, config, filter }));
  }

  toJSON() {
    const { id, type, config, filter } = this._value;

    return { id, type: type.toLocaleLowerCase(), config, filter };
  }

  static fromJSON(value: State) : Widget {
    const { id, type, config, filter } = value;
    const implementingClass = Widget.__registrations[type.toLocaleLowerCase()];

    if (implementingClass) {
      return implementingClass.fromJSON(value);
    }

    return new Widget(id, type, config, filter);
  }

  static registerSubtype(type: string, implementingClass: typeof Widget) {
    this.__registrations[type.toLocaleLowerCase()] = implementingClass;
  }
}

class Builder {
  value: Map<string, any>;
  constructor(value = Map()) {
    this.value = value;
  }

  id(value: string) {
    this.value = this.value.set('id', value);
    return this;
  }

  newId() {
    return this.id(uuid());
  }

  type(value: string) {
    this.value = this.value.set('type', value);
    return this;
  }

  config(value) {
    this.value = this.value.set('config', value);
    return this;
  }

  filter(value: string) {
    this.value = this.value.set('filter', value);
    return this;
  }

  build() : Widget {
    const { id, type, config, filter } = this.value.toObject();
    return new Widget(id, type, config, filter);
  }
}

Widget.Builder = Builder;

export default Widget;
