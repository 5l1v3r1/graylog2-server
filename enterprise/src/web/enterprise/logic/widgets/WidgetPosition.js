export default class WidgetPosition {
  constructor(col, row, height, width) {
    this._value = { col, row, height, width };
  }

  static fromJSON(value) {
    const { col, row, height, width } = value;
    const newCol = col === 'Infinity' ? Infinity : col;
    const newRow = row === 'Infinity' ? Infinity : row;
    const newHeight = height === 'Infinity' ? Infinity : height;
    const newWidth = width === 'Infinity' ? Infinity : width;

    return new WidgetPosition(newCol, newRow, newHeight, newWidth);
  }

  get col() {
    return this._value.col;
  }

  get row() {
    return this._value.row;
  }

  get height() {
    return this._value.height;
  }

  get width() {
    return this._value.width;
  }

  toJSON() {
    const { col, row, height, width } = this._value;

    const newCol = col === Infinity ? 'Infinity' : col;
    const newRow = row === Infinity ? 'Infinity' : row;
    const newHeight = height === Infinity ? 'Infinity' : height;
    const newWidth = width === Infinity ? 'Infinity' : width;

    return {
      col: newCol,
      row: newRow,
      height: newHeight,
      width: newWidth,
    };
  }
}
