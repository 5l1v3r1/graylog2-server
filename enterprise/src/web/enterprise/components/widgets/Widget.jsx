import React from 'react';
import PropTypes from 'prop-types';
import { MenuItem } from 'react-bootstrap';

import connect from 'stores/connect';
import UserNotification from 'util/UserNotification';
import { widgetDefinition } from 'enterprise/logic/Widget';
import { WidgetActions } from 'enterprise/stores/WidgetStore';
import { TitlesActions } from 'enterprise/stores/TitlesStore';
import { DashboardWidgetsActions } from 'enterprise/stores/DashboardWidgetsStore';

import WidgetFrame from './WidgetFrame';
import WidgetHeader from './WidgetHeader';
import WidgetFilterMenu from './WidgetFilterMenu';
import WidgetActionDropdown from './WidgetActionDropdown';
import WidgetHorizontalStretch from './WidgetHorizontalStretch';
import MeasureDimensions from './MeasureDimensions';

import styles from './Widget.css';
import EditWidgetFrame from './EditWidgetFrame';
import { ViewMetadataStore } from '../../stores/ViewMetadataStore';
import LoadingWidget from './LoadingWidget';
import ErrorWidget from './ErrorWidget';
import { WidgetErrorsMap } from './WidgetPropTypes';

class Widget extends React.Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
    widget: PropTypes.shape({
      computationTimeRange: PropTypes.object,
      config: PropTypes.object.isRequired,
    }).isRequired,
    data: PropTypes.any,
    errors: WidgetErrorsMap,
    height: PropTypes.number,
    width: PropTypes.number,
    fields: PropTypes.any.isRequired,
    onSizeChange: PropTypes.func.isRequired,
    onPositionsChange: PropTypes.func.isRequired,
    title: PropTypes.string.isRequired,
    position: PropTypes.object.isRequired,
    view: PropTypes.shape({
      activeQuery: PropTypes.string,
    }).isRequired,
  };

  static defaultProps = {
    height: 1,
    width: 1,
    data: undefined,
    errors: undefined,
  };

  static _visualizationForType(type) {
    return widgetDefinition(type).visualizationComponent;
  }

  static _editComponentForType(type) {
    return widgetDefinition(type).editComponent;
  }

  constructor(props) {
    super(props);
    this.state = {
      editing: false,
    };
  }

  _onDelete = (widget) => {
    // eslint-disable-next-line no-alert
    if (window.confirm(`Are you sure you want to remove the widget "${this.props.title}"?`)) {
      WidgetActions.remove(widget.id);
    }
  };

  _onDuplicate = (widgetId) => {
    WidgetActions.duplicate(widgetId);
  };

  _onToggleEdit = () => {
    this.setState(state => ({ editing: !state.editing }));
  };

  _onAddToOverview = (queryId, widgetId) => {
    DashboardWidgetsActions.addToDashboard(queryId, widgetId)
      .then(
        () => UserNotification.success('Added widget to dashboard.', 'Success!'),
        e => UserNotification.error(`Failed adding widget to dashboard: ${e}`, 'Error!'),
      );
  };

  _onWidgetConfigChange = (widgetId, config) => {
    WidgetActions.updateConfig(widgetId, config);
  };

  visualize = () => {
    const { data, errors } = this.props;
    if (errors && errors.length > 0) {
      return <ErrorWidget errors={errors} />;
    }
    if (data) {
      const { editing } = this.state;
      const { id, widget, height, width, fields } = this.props;
      const { config, computationTimeRange, filter } = widget;
      const VisComponent = Widget._visualizationForType(widget.type);
      return (<VisComponent id={id}
                    editing={editing}
                    title={widget.title}
                    config={config}
                    data={data}
                    fields={fields}
                    height={height}
                    width={width}
                    filter={filter}
                    onFinishEditing={this._onToggleEdit}
                    computationTimeRange={computationTimeRange} />);
    }
    return <LoadingWidget />;
  };
  render() {
    const { id, widget, fields, onSizeChange, title } = this.props;
    const { editing } = this.state;
    const { config, filter } = widget;
    const { activeQuery } = this.props.view;
    const visualization = this.visualize();
    const widgetActionDropdownCaret = <i className={`fa fa-chevron-down ${styles.widgetActionDropdownCaret} ${styles.tonedDown}`} />;
    if (editing) {
      let editWidgetFrameContent = null;
      const EditComponent = Widget._editComponentForType(widget.type);
      return (
        <EditWidgetFrame widgetId={id}>
          <span ref={(elem) => { editWidgetFrameContent = elem; }}>
            <MeasureDimensions>
              <WidgetHeader title={title}
                            onRename={newTitle => TitlesActions.set('widget', id, newTitle)}
                            editing={editing}>
                <WidgetFilterMenu onChange={newFilter => WidgetActions.filter(id, newFilter)} value={filter}>
                  <i className={`fa fa-filter ${styles.widgetActionDropdownCaret} ${filter ? styles.filterSet : styles.filterNotSet}`} />
                </WidgetFilterMenu>
                {' '}
                <WidgetActionDropdown element={widgetActionDropdownCaret} container={() => editWidgetFrameContent}>
                  <MenuItem onSelect={this._onToggleEdit}>Finish Editing</MenuItem>
                </WidgetActionDropdown>
              </WidgetHeader>
              <EditComponent config={config}
                             fields={fields}
                             id={id}
                             onChange={newWidgetConfig => this._onWidgetConfigChange(id, newWidgetConfig)}>
                {visualization}
              </EditComponent>
            </MeasureDimensions>
          </span>
        </EditWidgetFrame>
      );
    }
    let container = null;
    return (
      <WidgetFrame widgetId={id} onSizeChange={onSizeChange}>
        <span ref={(elem) => { container = elem; }}>
          <MeasureDimensions>
            <WidgetHeader title={title}
                          onRename={newTitle => TitlesActions.set('widget', id, newTitle)}
                          editing={editing}>
              <WidgetHorizontalStretch widgetId={widget.id}
                                       widgetType={widget.type}
                                       onStretch={this.props.onPositionsChange}
                                       position={this.props.position} />
              {' '}
              <WidgetFilterMenu onChange={newFilter => WidgetActions.filter(id, newFilter)} value={filter}>
                <i className={`fa fa-filter ${styles.widgetActionDropdownCaret} ${filter ? styles.filterSet : styles.filterNotSet}`} />
              </WidgetFilterMenu>
              {' '}
              <WidgetActionDropdown element={widgetActionDropdownCaret} container={() => container}>
                <MenuItem onSelect={this._onToggleEdit}>Edit</MenuItem>
                <MenuItem onSelect={() => this._onDuplicate(id)}>Duplicate</MenuItem>
                <MenuItem divider />
                <MenuItem onSelect={() => this._onAddToOverview(activeQuery, id)}>Add to overview</MenuItem>
                <MenuItem divider />
                <MenuItem onSelect={() => this._onDelete(widget)}>Delete</MenuItem>
              </WidgetActionDropdown>
            </WidgetHeader>
            {visualization}
          </MeasureDimensions>
        </span>
      </WidgetFrame>
    );
  }
}

export default connect(Widget, { view: ViewMetadataStore });
