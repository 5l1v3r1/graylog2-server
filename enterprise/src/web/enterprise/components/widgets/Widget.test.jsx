import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import mockComponent from 'helpers/mocking/MockComponent';
import Widget from './Widget';
import WidgetPosition from '../../logic/widgets/WidgetPosition';
import LoadingWidget from './LoadingWidget';

jest.mock('../searchbar/QueryInput', () => mockComponent('QueryInput'));
jest.mock('./WidgetHeader', () => mockComponent('WidgetHeader'));
jest.mock('graylog-web-plugin/plugin', () => ({
  PluginStore: {
    exports: () => ([
      {
        type: 'dummy',
        displayName: 'Some Dummy Visualization',
        visualizationComponent: mockComponent('DummyVisualization'),
      },
    ]),
  },
}));

const createNodeMock = (element) => {
  if (element.type === 'div') {
    return {
      clientHeight: 420,
      clientWidth: 420,
    };
  }
  return null;
};

describe('<Widget />', () => {
  it('should render with empty props', () => {
    const options = { createNodeMock };
    const wrapper = renderer.create((
      <Widget widget={{}}
              onSizeChange={() => {}}
              position={new WidgetPosition(1, 1, 1, 1)} />
    ), options);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
  it('should render loading widget for widget without data', () => {
    const wrapper = mount((
      <Widget widget={{}}
              onSizeChange={() => {}}
              position={new WidgetPosition(1, 1, 1, 1)} />
    ));
    expect(wrapper.find(LoadingWidget)).toHaveLength(1);
  });
  it('should render correct widget visualization for widget with data', () => {
    const wrapper = mount((
      <Widget widget={{
        type: 'dummy',
      }}
              onSizeChange={() => {}}
              data={[]}
              position={new WidgetPosition(1, 1, 1, 1)} />
    ));
    expect(wrapper.find(LoadingWidget)).toHaveLength(0);
    expect(wrapper.find('DummyVisualization')).toHaveLength(1);
  });
});
