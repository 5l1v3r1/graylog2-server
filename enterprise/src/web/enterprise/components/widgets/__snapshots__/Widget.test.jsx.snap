// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Widget /> should render with empty props 1`] = `
<div
  className="widget"
  style={
    Object {
      "overflow": "hidden",
    }
  }
>
  <div
    style={
      Object {
        "height": "95%",
        "padding": "5px",
      }
    }
  >
    <span>
      <span
        className="WidgetHeader"
      >
        <span
          editing={false}
          onRename={[Function]}
        >
          <span>
            <i
              className="fa fa-arrows-h button"
              onClick={[Function]}
              role="link"
              tabIndex={0}
            />
          </span>
           
          <i
            className="fa fa-filter widgetActionDropdownCaret filterNotSet"
            onClick={[Function]}
          />
           
          <span>
            <span
              onClick={[Function]}
              role="presentation"
            >
              <span>
                <i
                  className="fa fa-chevron-down widgetActionDropdownCaret tonedDown"
                />
              </span>
            </span>
          </span>
        </span>
      </span>
      <div
        className="spinnerContainer"
      >
        <i
          className="fa fa-spin fa-3x fa-refresh spinner"
        />
      </div>
    </span>
  </div>
</div>
`;
